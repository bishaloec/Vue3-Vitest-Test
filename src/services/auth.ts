// src/services/auth.ts
export interface LoginResult {
    success: boolean;
    message?: string;
    token?: string;
  }
  
  export function useAuthService() {
    const login = async (email: string, password: string): Promise<LoginResult> => {
      // In a real app, this would make an API request
      // This is a simplified mock implementation
      await new Promise(resolve => setTimeout(resolve, 500)); // Simulate API delay
      
      // Mock authentication logic
      if (email === 'test@example.com' && password === 'Password123') {
        return {
          success: true,
          token: 'mock-jwt-token'
        };
      }
      
      return {
        success: false,
        message: 'ログインメールとパスワードは一致しません'
      };
    };
    
    return {
      login
    };
  }
  