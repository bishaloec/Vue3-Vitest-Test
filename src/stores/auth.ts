import { defineStore } from 'pinia'

export const useAuthStore = defineStore('auth', {
  state: () => ({
    isAuthenticated: false,
    user: null as string | null
  }),
  actions: {
    login(email: string) {
      this.isAuthenticated = true
      this.user = email // ユーザー名を正しく設定
    },
    logout() {
      this.isAuthenticated = false
      this.user = null
    }
  }
})