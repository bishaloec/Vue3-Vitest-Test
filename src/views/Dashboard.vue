<template>
    <div class="dashboard-container">
      <h1>ようこそ、{{ user }} さん！</h1>
      <button @click="handleLogout" class="logout-button">ログアウト</button>
    </div>
  </template>
  
  <script lang="ts">
  import { defineComponent, computed } from 'vue'
  import { useRouter } from 'vue-router'
  import { useAuthStore } from '../stores/auth.ts'
  
  export default defineComponent({
    setup() {
      const authStore = useAuthStore()
      const router = useRouter()
  
      const user = computed(() => authStore.user)
  
      const handleLogout = () => {
        authStore.logout()
        router.push('/login')
      }
  
      return {
        user,
        handleLogout
      }
    }
  })
  </script>
  
  <style scoped>
  .dashboard-container {
    max-width: 800px;
    margin: 50px auto;
    padding: 30px;
    text-align: center;
    position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  }
  
  .logout-button {
    padding: 12px 25px;
    background-color: #dc3545;
    color: white;
    border: none;
    border-radius: 5px;
    cursor: pointer;
    transition: background-color 0.3s;
  }
  
  .logout-button:hover {
    background-color: #c82333;
  }
  </style>