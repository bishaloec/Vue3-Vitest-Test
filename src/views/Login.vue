<template>
  <div class="login-container">
    <h2>ログイン「テスト」</h2>
    
    <div v-if="successMessage" class="success-message">
      {{ successMessage }}
    </div>
    
    <form @submit.prevent="handleLogin" class="login-form">
      <div class="form-group">
        <label for="email">📧 メールアドレス :</label>
        <input 
          id="email"
          v-model="email"
          type="email"
          placeholder="メールアドレスを入力してください"
          required
          @blur="validateEmail"
        />
        <p v-if="emailError" class="error-message">{{ emailError }}</p>
      </div>
      
      <div class="form-group">
        <label for="password">🔒 パスワード:</label>
        <input 
          id="password"
          v-model="password"
          type="password"
          placeholder="パスワードを入力してください"
          required
          @blur="validatePassword"
        />
        <p v-if="passwordError" class="error-message">{{ passwordError }}</p>
        <p v-if="passwordWarning" class="warning-message">{{ passwordWarning }}</p>
      </div>
      
      <div v-if="loginError" class="error-message">
        {{ loginError }}
      </div>
      
      <button 
        type="submit" 
        :disabled="isLoginDisabled || isLoading"
        class="login-button"
      >
        <span v-if="isLoading">ログイン中...</span>
        <span v-else>ログイン</span>
      </button>
    </form>
  </div>
</template>

<script lang="ts">
import { defineComponent, ref, computed } from 'vue'
import { useAuthService } from '../services/auth'
import { useAuthStore } from '../stores/auth.ts'
import { useRouter } from 'vue-router'


export default defineComponent({
  name: 'LoginComponent',
  setup() {
    const email = ref('')
    const password = ref('')
    const emailError = ref('')
    const passwordError = ref('')
    const passwordWarning = ref('')
    const loginError = ref('')
    const successMessage = ref('')
    const isLoading = ref(false)
    
    const authService = useAuthService()

    const router = useRouter()
  
    const validateEmail = () => {
      emailError.value = ''
      
      if (!email.value) {
        emailError.value = 'メールアドレスが必要'
        return false
      }
      
      const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
      if (!emailRegex.test(email.value)) {
        emailError.value = '有効なメールアドレスを入力してください'
        return false
      }
      
      return true
    }
    
    const validatePassword = () => {
      passwordError.value = ''
      passwordWarning.value = ''
      
      if (!password.value) {
        passwordError.value = 'パスワードが必要'
        return false
      }
      
      // パスワードの強さを確認
      if (password.value.length < 8) {
        passwordWarning.value = 'パスワードは少なくとも8文字以上の長さが必要です'
      } else if (!/[A-Z]/.test(password.value)) {
        passwordWarning.value = 'パスワードには、少なくとも1つの大文字が含まれている必要があります'
      } else if (!/[0-9]/.test(password.value)) {
        passwordWarning.value = 'パスワードには少なくとも1つの番号が含まれている必要があります'
      }
      
      return true
    }
    
    const isLoginDisabled = computed(() => {
      return !email.value || !password.value || !!emailError.value || !!passwordError.value
    })
    
    const handleLogin = async () => {
      // 以前のエラーをリセット
      loginError.value = ''
      successMessage.value = ''
      
      // フォームを検証
      const isEmailValid = validateEmail()
      const isPasswordValid = validatePassword()
      
      if (!isEmailValid || !isPasswordValid) {
        return
      }
      
      try {
        isLoading.value = true
        const result = await authService.login(email.value, password.value)
        
        if (result.success) {
          successMessage.value = 'ログインが成功'
          const authStore = useAuthStore()
          authStore.login(email.value)
          router.push('/dashboard') // Vue Routerを使用
          // 実際のアプリでは、ここでリダイレクトするか、auth stateを更新する
        } else {
          loginError.value = result.message || 'Login failed. Please check your credentials.'
        }
      } catch (error) {
        loginError.value = 'An unexpected error occurred. Please try again.'
        console.error('Login error:', error)
      } finally {
        isLoading.value = false
      }
    }
    
    return {
      email,
      password,
      emailError,
      passwordError,
      passwordWarning,
      loginError,
      successMessage,
      isLoading,
      isLoginDisabled,
      validateEmail,
      validatePassword,
      handleLogin
    }
  }
})
</script>

<style scoped>
/* Center the login container on the screen */
.login-container {
  max-width: 450px;
  width: 100%;
  padding: 20px;
  background: #fff;
  border-radius: 8px;
  box-shadow: 0px 4px 10px rgba(0, 0, 0, 0.1);
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.login-form {
  display: flex;
  flex-direction: column;
  gap: 20px;
  width: 100%;
}

.form-group {
  display: flex;
  flex-direction: column;
  gap: 8px;
}

input {
  padding: 10px;
  border: 1px solid #ccc;
  border-radius: 4px;
}

.error-message {
  color: #d9534f;
  font-size: 14px;
  margin-top: 5px;
}

.warning-message {
  color: #f0ad4e;
  font-size: 14px;
  margin-top: 5px;
}

.success-message {
  color: #5cb85c;
  margin-bottom: 20px;
  padding: 10px;
  background-color: #dff0d8;
  border-radius: 4px;
}

.login-button {
  padding: 10px 15px;
  background-color: #337ab7;
  color: white;
  border: none;
  border-radius: 4px;
  cursor: pointer;
}

.login-button:disabled {
  background-color: #cccccc;
  cursor: not-allowed;
}
</style>