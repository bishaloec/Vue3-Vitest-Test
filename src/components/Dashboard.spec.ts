import { describe, it, expect, beforeEach } from 'vitest';
import { mount } from '@vue/test-utils';
import { createRouter, createMemoryHistory } from 'vue-router';
import Dashboard from '@/views/Dashboard.vue';  
import Login from '@/views/Login.vue';          
import { createPinia, setActivePinia } from 'pinia';
import { useAuthStore } from '../stores/auth.ts'


// Vue RouterをMock
const routes = [
  { path: '/', redirect: '/login' },
  { path: '/dashboard', component: Dashboard , meta: { requiresAuth: true }},
  { path: '/login', component: Login },
];

const router = createRouter({
  history: createMemoryHistory(),
  routes,
});

describe('DashboardComponent', () => {
  let wrapper: any

  beforeEach(() => {
     setActivePinia(createPinia()); // Initialize Pinia if you're using it
    // 認証状態を設定（ログイン状態をシミュレート）
    const authStore = useAuthStore()
    authStore.login('test@example.com') // ユーザー名を設定
     // Dashboardコンポーネントをマウント
     wrapper = mount(Dashboard, {
        global: {
          plugins: [router], // Provide router to the test
        },
      });
  
  });

  it('ログインが成功した後、ダッシュボードにリダイレクトする', async () => {
    router.push('/login'); // ログインページへ繊維
    await router.isReady();

    // dashboardページへ繊維
    await router.push('/dashboard');

    expect(router.currentRoute.value.path).toBe('/dashboard');

    expect(wrapper.find('h1').text()).toBe('ようこそ、test@example.com さん！');
  });

  it('ログアウトがクリックされた時にログインにリダイレクトする', async () => {
    router.push('/dashboard');
    await router.isReady();

    // [ログアウト]ボタンをクリックする
    await wrapper.find('button').trigger('click')

    // モックAPI呼び出しが解決するのを待ち
    await new Promise(resolve => setTimeout(resolve, 20))

    // ログインしてログインしてログインしていると仮定する
    expect(router.currentRoute.value.path).toBe('/login');
  });

  it('ダッシュボード snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
