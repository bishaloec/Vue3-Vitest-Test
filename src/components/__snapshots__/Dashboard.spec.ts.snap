// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DashboardComponent > ダッシュボード snapshot 1`] = `
"<div data-v-7f773d42="" class="dashboard-container">
  <h1 data-v-7f773d42="">ようこそ、test@example.com さん！</h1><button data-v-7f773d42="" class="logout-button">ログアウト</button>
</div>"
`;
